### PR TITLE
Refactoring: DAG migrated to Plexus's Digraph

### PR DESCRIPTION
**Pull request by prathamesh-mutkure**
## Which problem is this PR solving?
- This issue is related to LFX task #1466 
- I can create a more specific sub-issue if that's more suitable

## Description of the changes
- I've removed `cytoscape` and replaced it with `Digraph` from `Plexus`
- Had to add support for labels in `Digraph` to support this change

## How was this change tested?
- The changes were tested visually
- Tested with test data involving around 50 nodes

Before - 

Screenshot 2023-11-18 at 22 06 10

Now -

Screenshot 2023-11-18 at 22 06 22

Tha additional tab visible in above images has been removed

## Checklist
- [x] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [x] I have signed all commits
- [ ] I have added unit tests for the new functionality
- [x] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `yarn lint` and `yarn test`

### CODE REVIEW

**Comment by prathamesh-mutkure** [comment]
Hey @yurishkuro, the two test cases will fail for `DAG`, I'll fix them ASAP, you can review the changes in the meantime

**Comment by prathamesh-mutkure** [comment]
Hey @yashrsharma44, tagging you for review

The code coverage has dropped, will try to increase it

File: packages/jaeger-ui/package.json [hunk 1]
```diff
@@ -55,9 +55,6 @@
     "classnames": "^2.2.5",
     "combokeys": "^3.0.0",
     "copy-to-clipboard": "^3.1.0",
-    "cytoscape": "^3.2.1",
-    "cytoscape-dagre": "^2.0.0",
-    "dagre": "^0.8.5",
```
Comment (yurishkuro) [inline packages/jaeger-ui/package.json:60]: wow, really, we're even removing dagre?

File: packages/jaeger-ui/src/components/DependencyGraph/DAG.test.js [hunk 2]
```diff
@@ -13,7 +13,7 @@
 // limitations under the License.
 
 import React from 'react';
-import { render, screen } from '@testing-library/react';
+import { shallow } from 'enzyme';
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/DependencyGraph/DAG.test.js:16]: we are trying to move away from enzyme, please do not reintroduce it here, use testing-library

Comment (yurishkuro) [reply packages/jaeger-ui/src/components/DependencyGraph/DAG.test.js:16]: especially considering that the tests here are extremely primitive, just a check for top-level element

File: packages/jaeger-ui/src/components/DependencyGraph/DAG.tsx [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+  }[];
+};
+
+type TState = {
+  nodes: TVertex[];
+  edges: TEdge[];
+};
+
+const renderNode = (vertex: TVertex<TVertex>): ReactNode => {
+  return (
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/DependencyGraph/DAG.tsx:37]: how is it possible for this function to never be called in the test?

File: packages/jaeger-ui/src/components/DependencyGraph/DAG.tsx [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    <div className="DAG--node">
+      <div className="DAG--nodeCircle" />
+
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/DependencyGraph/DAG.tsx:40]:
```suggestion
```

File: packages/plexus/src/Digraph/SvgEdge.tsx [hunk 5]
```diff
@@ -62,15 +64,31 @@ export default class SvgEdge<U = {}> extends React.PureComponent<TProps<U>> {
       },
       getProps(setOnEdge, layoutEdge, renderUtils)
     );
+
+    const [startX, startY] = pathPoints[0];
+    const [endX, endY] = pathPoints[pathPoints.length - 1];
+
+    const xOffset = (label?.length ?? 0) * 5;
+    const labelX = (startX + endX) / 2;
+    const labelY = (startY + endY) / 2;
```
Comment (yurishkuro) [inline packages/plexus/src/Digraph/SvgEdge.tsx:73]:
```suggestion
    const computeLabelCoords = () => {
      const [startX, startY] = pathPoints[0];
      const [endX, endY] = pathPoints[pathPoints.length - 1];
      const xOffset = (label?.length ?? 0) * 5;
      const labelX = (startX + endX) / 2 - xOffset;
      const labelY = (startY + endY) / 2;
      return { labelX, labelY };
    };
    const { labelX, labelY } = computeLabelCoords();
```
this way you do not pollute the scope with temporary variables that are only relevant for calculation

File: packages/plexus/src/Digraph/SvgEdge.tsx [hunk 6]
```diff
@@ -62,15 +64,31 @@ export default class SvgEdge<U = {}> extends React.PureComponent<TProps<U>> {
...
+          d={d}
+          fill="none"
+          vectorEffect="non-scaling-stroke"
+          markerEnd={markerEnd}
+          markerStart={markerStart}
+          {...customProps}
+        />
+
+        {label && (
+          <text x={labelX - xOffset} y={labelY} fill="#000" fontSize="1rem" fontWeight="bold">
```
Comment (yurishkuro) [inline packages/plexus/src/Digraph/SvgEdge.tsx:87]:
```suggestion
          <text x={labelX} y={labelY} fill="#000" fontSize="1rem" fontWeight="bold">
```

File: packages/jaeger-ui/src/components/DependencyGraph/DAG.test.js [hunk 7]
```diff
@@ -63,8 +64,10 @@ describe('<DAG>', () => {
         parent: 'parent-id',
       },
     ];
-    render(<DAG serviceCalls={serviceCalls} />);
-    expect(screen.getByTestId('cy')).toBeDefined();
+
+    const wrapper = shallow(<DAG serviceCalls={serviceCalls} />);
+
+    expect(wrapper.exists('.DAG')).toBe(true);
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/DependencyGraph/DAG.test.js:70]: can we do something smarter, like count how many nodes and edges are displayed?

**Comment by yurishkuro** [comment]
* ping zoom does not seem to work at all, I am getting the whole browser window changing zoom instead of the graph div only, as it was in the old DAG. I know that zoom does work in DDG (but does not seem to work in Trace Diff)
* is dragging of the nodes not supported by Plexus at all? Not the end of the world if it's not.

**Comment by yurishkuro** [comment]
Another odd thing: if I tab to Force Directed Graph and then back to DAG, the minimap goes rather nuts - the light-gray box is supposed to indicate current zoom view in the viewport, and it shows outside of it, even though the main display is fine. Clicking on the reset button returns the box back into full size minimap. I suspect it may be related to the zoom nor working.

image

**Comment by yurishkuro** [comment]
I tried this with graphs with larger number of nodes. With 15 nodes it looks like this:
image
Curious why the nodes / text are so small, there's enough space between them for larger nodes.

The zoom seems to work with these larger graphs.

File: packages/jaeger-ui/package.json (hunk 1 above)
Comment (prathamesh-mutkure) [reply packages/jaeger-ui/package.json:60]: Yes, not used anywhere anymore

File: packages/jaeger-ui/src/components/DependencyGraph/DAG.test.js (hunk 2 above)
Comment (prathamesh-mutkure) [reply packages/jaeger-ui/src/components/DependencyGraph/DAG.test.js:16]: There's a comment in the test file that we don't really care about the rendered canvas actually, we still can check for the number of nodes and edges to be more precise

File: packages/jaeger-ui/src/components/DependencyGraph/DAG.tsx (hunk 3 above)
Comment (prathamesh-mutkure) [reply packages/jaeger-ui/src/components/DependencyGraph/DAG.tsx:37]: Yeah, this is what caused test coverage to decrease

File: packages/plexus/src/Digraph/SvgEdge.tsx (hunk 5 above)
Comment (prathamesh-mutkure) [reply packages/plexus/src/Digraph/SvgEdge.tsx:73]: That's a good suggestion, more memory efficient

File: packages/jaeger-ui/src/components/DependencyGraph/DAG.test.js (hunk 7 above)
Comment (prathamesh-mutkure) [reply packages/jaeger-ui/src/components/DependencyGraph/DAG.test.js:70]: Yes, that will be better

**Comment by prathamesh-mutkure** [comment]
> * ping zoom does not seem to work at all, I am getting the whole browser window changing zoom instead of the graph div only, as it was in the old DAG. I know that zoom does work in DDG (but does not seem to work in Trace Diff)

The graph is likely smaller, you seem to got the zoom working with larger graph

> is dragging of the nodes not supported by Plexus at all? Not the end of the world if it's not.

Did not see anything in the documentation and code sadly

> if I tab to Force Directed Graph and then back to DAG, the minimap goes rather nuts

I'll check this one and try to fix it

> I tried this with graphs with a larger number of nodes. With 15 nodes it looks like this:
> Curious why the nodes / text are so small, there's enough space between them for larger nodes

The coordinates are actually returned by Plexus, the size of nodes is static, and we can play around with the size of nodes

Also, can you share data for those 15 nodes?

**Comment by yurishkuro** [comment]
15-node graph https://gist.github.com/yurishkuro/1f7dfbceebc5f192a2e3108c53ad96da

**Comment by yurishkuro** [comment]
> The graph is likely smaller, you seem to got the zoom working with larger graph

that doesn't explain why the zoom doesn't work - with the larger 15-node graph I can zoom it so that the nodes are twice as large as in the default hotrod graph (which doesn't zoom)

File: packages/jaeger-ui/src/components/DependencyGraph/DAG.test.js [hunk 8]
```diff
@@ -13,8 +13,10 @@
 // limitations under the License.
 
 import React from 'react';
+import ShallowRenderer from 'react-test-renderer/shallow';
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/DependencyGraph/DAG.test.js:16]: why do we need to use `ShallowRenderer` and not the standard `render`? Does `ShallowRenderer` actually render _more_ details?

**Review by yurishkuro — APPROVED** [review]

**Pull request merged**